src/kdbindings/signal.h

```diff
@@ -260,6 +260,37 @@ class Signal
         return ConnectionHandle{ m_impl, m_impl->connect(slot) };
     }
 
+    /**
+     * Establishes a connection between a signal and a slot, allowing the slot to access its own connection handle.
+     * This method is particularly useful for creating connections that can manage themselves, such as disconnecting
+     * after being triggered a certain number of times or under specific conditions. It wraps the given slot function
+     * to include a shared pointer to the ConnectionHandle as the first parameter, enabling the slot to interact with
+     * its own connection state directly.
+     *
+     * @param slot A std::function that takes a shared_ptr<ConnectionHandle> followed by the signal's parameter types.
+     * @return A shared_ptr to the ConnectionHandle associated with this connection, allowing for advanced connection management.
+     */
+    std::shared_ptr<ConnectionHandle> connectReflective(std::function<void(std::shared_ptr<ConnectionHandle>, Args...)> const &slot)
```

Comment:
I don't think we need the shared_ptr here.

The `connectDeferred` method already pretty much does reflective connection.
Why can't we use the same mechanism here?

Basically just rename `Connection::slotDeferred` to `Connection::slotReflective`.

Reply:
I'm not sure, but what's the need for `Connection::SlotReflective` ? Currently we are using `connect` to call with the  `wrappedSlot` , is there any deal with `ConnectionEvaluator` for the single-shot mechanism? If not I don't think if we want it through the `connectDeferred` , and I think we can remove the `shared_ptr` from here,

Reply:
What I'm trying to say is that both single-shot connections and `connectDeferred` both end up using a reflective connection.

This is already the case with our current `connectDeferred` implementation. It is *already* reflective.

When you call `connectDeferred`, what ends up being connected to the signal is a function that takes its own ConnectionHandle as argument. That is basically a reflective connection.
So we should be able to use this same, existing mechanism, instead of having to reinvent it.

See: https://github.com/KDAB/KDBindings/blob/7a2b6d2e4fcfbe112d92c4a3c63d6f1de7299e2a/src/kdbindings/signal.h#L99
The lambda that's created there takes:
1. A ConnectionHandle reference
2. The arguments emitted by the Signal

That is already a reflective connection.

Reply:
Sorry If I'm still misunderstanding it, So do we want to have an another method `connectReflective` or not? I think we need it for accepting the slot with `ConnectionHandle` parameter, and then we want to wrap that slot and pass it to the `connectDeferred` method, right?  

But in case using the `connectDeferred` directly for single-shot, still for that we have to change the definition of `connectDeferred` to accept slot of type `std::function<void(ConnectionHandle&, Args...)> slot` , I'm not sure if it's worth to change the definition as we have to change the tests as well and even this method is not heavily used but still it would be 'breaking' change. We might wanna use type_traits for the `connectDeferred` method to accept slots both with and without a `ConnectionHandle` parameter for keeping the backward compatibility. Also do we want to make the `ConnectionEvaluator` optional, if we want to update the existing mechanism for single-shot?

Basically from the user perspective this should be the case right for the above case? 

```
    auto handle = mySignal.connectDeferred(evaluator may be optional, [&val](ConnectionHandle& selfHandle, int value) {
            val += value;

            selfHandle.disconnect();
        });
```

Reply:
Ah, that's not what I meant!

Let me try explaining it differently :)

I'm totally onboard with adding `connectReflective` as public API.
What I just mean is that `connectDeferred` *internally* already uses this!

Basically this code that we already have:
```cpp
auto weakEvaluator = std::weak_ptr<ConnectionEvaluator>(evaluator);

//                               This is a reflective connection
// vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
auto deferredSlot = [weakEvaluator = std::move(weakEvaluator), slot](const ConnectionHandle &handle, Args... args) {
    if (auto evaluatorPtr = weakEvaluator.lock()) {
        auto lambda = [slot, args...]() {
            slot(args...);
        };
        evaluatorPtr->enqueueSlotInvocation(handle, lambda);
    } else {
        throw std::runtime_error("ConnectionEvaluator is no longer alive");
    }
};
// ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Connection newConnection;
newConnection.m_connectionEvaluator = evaluator;
newConnection.slotDeferred = deferredSlot;
//            ^^^^^^^^^^^^ this could just be called slotReflective - that's basically what it is already!

return m_connections.insert(std::move(newConnection));
```
Is already doing a reflective connection.

The `deferredSlot` lambda is exactly the signature that we expect from a reflective connection:
`(const ConnectionHandle &handle, Args... args)`

So we clearly **already have** reflective connections internally.
Why not reuse that same mechanism?

The `slotDeferred` method therefore simply **uses** the reflective connections internally.
That's something the user doesn't and shouldn't need to know about, but it's still something we can reuse internally.

We'll still definitely want to provide a separate `connectReflective` method as public API, but why does that new method have to reinvent the wheel? We already have the mechanism, let's just expose it.

Reply:
Right, of course, it needs to be given a non-const reference, but that should be easy enough to do :)

Reply:
@phyBrackets any news on this?

I'll not be available for the next two weeks, so if you want to get another round of reviews in, your changes need to be uploaded before Friday of this week.